src/modules/flow/oauth/oauth.c

```diff
@@ -0,0 +1,596 @@
+        sol_http_client_connection_cancel(connection);
+    sol_ptr_vector_clear(&mdata->pending_conns);
+}
+
+static int
+v1_parse_response(struct v1_request_data *req_data, const struct sol_http_response *response)
+{
+    int r = -EINVAL;
+    char *token, *token_secret, *url = NULL;
+    struct sol_http_response start_response = {
+        .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+        .content = SOL_BUFFER_INIT_EMPTY,
+        .param = SOL_HTTP_REQUEST_PARAM_INIT,
+        .response_code = SOL_HTTP_STATUS_FOUND,
+    };
+
+    token = strstr(response->content.data, "oauth_token=");
+    token_secret = strstr(response->content.data, "oauth_token_secret=");
+
+    if (token_secret) {
+        token = strndup(token, token_secret - token - 1);
+        SOL_NULL_CHECK(token, -ENOMEM);
+
+        r = asprintf(&url, "%s?%s", req_data->mdata->authorize_token_url, token);
+        SOL_INT_CHECK_GOTO(r, < 0, err);
+        if (!sol_http_param_add(&start_response.param,
+            SOL_HTTP_REQUEST_PARAM_HEADER("Location", url)))
+            goto err_param;
+
+        start_response.url = url;
+
+        r = sol_http_server_send_response(req_data->request, &start_response);
+        SOL_INT_CHECK_GOTO(r, < 0, err_send);
+    }
+
+err_send:
+    sol_http_param_free(&start_response.param);
+err_param:
+    free(url);
+err:
+    free(token);
+    return r;
+}
+
+static void
+v1_request_finished(void *data,
+    const struct sol_http_client_connection *connection,
+    struct sol_http_response *response)
+{
+    int r;
+    struct v1_request_data *req_data = data;
+    struct v1_data *mdata = req_data->mdata;
+
+    if (sol_ptr_vector_remove(&mdata->pending_conns, connection) < 0)
+        SOL_WRN("Failed to find pending connection %p", connection);
+
+    SOL_NULL_CHECK_GOTO(response, err);
+    SOL_HTTP_RESPONSE_CHECK_API_GOTO(response, err);
+    SOL_INT_CHECK_GOTO(response->content.used, == 0, err);
+
+    if (response->response_code != SOL_HTTP_STATUS_OK) {
+        SOL_WRN("Response from %s - %d", response->url, response->response_code);
+        goto err;
+    }
+
+    r = v1_parse_response(req_data, response);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    goto end;
+
+err:
+    r = sol_http_server_send_response(req_data->request, &((struct sol_http_response) {
+            .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+            .content = SOL_BUFFER_INIT_EMPTY,
+            .param = SOL_HTTP_REQUEST_PARAM_INIT,
+            .response_code = SOL_HTTP_STATUS_INTERNAL_SERVER_ERROR
+        }));
+    if (r < 0) {
+        SOL_WRN("Could not send fail response for %s",
+            sol_http_request_get_url(req_data->request));
+
+    }
+end:
+    free(req_data->timestamp);
+    free(req_data->nonce);
+    free(req_data);
+}
+
+static char *
+get_url_callback(const char *namespace)
+{
+    char *temp, *url;
+    int r;
+
+    r = asprintf(&temp, "http://localhost:%d/%s/oauth_callback",
```

Comment:
How is this supposed to work? I mean -- it'll work fine on your machine, but how is this going to work if, say, Soletta is running on a board, and I'm going through OAuth steps there with my cell phone?